bokehjs/src/coffee/document.coffee

```diff
@@ -614,6 +614,10 @@ class Document
         new_references[id] = value
 
     Document._initialize_references_json(references_json, old_references, new_references)
+    for id, model of new_references
+      model.attach_document(@)
+      if model.layoutable is true
+        @_add_layoutable(model)
```

Comment:
and then of course if the children of a container change, the container needs to notify its parent, which may be the document if the container is a root, that constraints have changed.